kimchi/src/circuits/lookup/index.rs

```diff
@@ -0,0 +1,197 @@
+    #[serde(bound = "LookupConfiguration<F>: Serialize + DeserializeOwned")]
+    pub configuration: LookupConfiguration<F>,
+}
+
+impl<F: FftField + SquareRootField> LookupConstraintSystem<F> {
+    pub fn create(
+        gates: &[CircuitGate<F>],
+        lookup_tables: Vec<LookupTable<F>>,
+        domain: &EvaluationDomains<F>,
+    ) -> Result<Option<Self>, LookupError> {
+        let lookup_info = LookupInfo::<F>::create();
+
+        //~ 1. If no lookup is used in the circuit, do not create a lookup index
+        match lookup_info.lookup_used(gates) {
+            None => Ok(None),
+            Some(lookup_used) => {
+                let d1_size = domain.d1.size();
+
+                // The maximum number of entries that can be provided across all tables.
+                // Since we do not assert the lookup constraint on the final `ZK_ROWS` rows, and
+                // because the row before is used to assert that the lookup argument's final
+                // product is 1, we cannot use those rows to store any values.
+                let max_num_entries = d1_size - (ZK_ROWS as usize) - 1;
+
+                //~ 2. Get the lookup selectors and lookup tables (TODO: how?)
+                let (lookup_selectors, gate_lookup_tables) =
+                    lookup_info.selector_polynomials_and_tables(domain, gates);
+
+                //~ 3. Concatenate runtime lookup tables with the ones used by gates
+                let lookup_tables: Vec<_> = gate_lookup_tables
+                    .into_iter()
+                    .chain(lookup_tables.into_iter())
+                    .collect();
+
+                //~ 4. Get the highest number of columns `max_table_width`
+                //~    that a lookup table can have.
+                let max_table_width = lookup_tables
+                    .iter()
+                    .map(|table| table.data.len())
+                    .max()
+                    .unwrap_or(0);
+
+                //~ 5. Add the table ID stuff
+                let mut lookup_table = vec![Vec::with_capacity(d1_size); max_table_width];
+                let mut table_ids: Vec<F> = Vec::with_capacity(d1_size);
+                let mut non_zero_table_id = false;
+                //~ 6. For each table:
+                for table in lookup_tables.iter() {
+                    let table_len = table.data[0].len();
+
+                    //~ b. Make sure that if table with id 0 is used, then it's the XOR table.
+                    //~    We do this because we use a table with id 0 and
+                    //~
+                    if table.id == 0 {
+                        if !table.has_zero_entry() {
+                            return Err(LookupError::TableIDZeroMustHaveZeroEntry);
+                        }
+                    }
+
+                    // Update table IDs
+                    if table.id != 0 {
+                        non_zero_table_id = true;
+                    }
+                    //~ c. Update table IDs
```

Comment:
Nit: `b`

Reply:
This seems under-detailed for a spec. Shouldn't this be something like 'extend the table IDs column with the table ID for this table, repeated for each entry in the table' or some such?